### PR TITLE
Refactor runtime to use context and new runtime api

### PR DESCRIPTION
**Pull request by firstdoit**
- [x] Use static class properties when possible
- [x] No more messy globals running around
- [x] No more mutating global state unnecessarily
- [x] Each nested extension point gets it's `params` via props
- [x] Implement componentDidCatch on ExtensionPoint
- [x] Use `eventemitter3` and reference it through context
- [x] Assets are loaded on demand when extension points render (support for async extension points!)
- [x] Update subscriptions if component changes
- [x] Mount ApolloProvider inside RenderProvider
- [x] Pass prefetchPage in context
- [x] `extension.component` might be an array indicating nested components
- [x] lint codebase

### CODE REVIEW

File: react/utils/assets.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
+function isRelative(path) {
+  return path[0] !== '/'
+}
+
+function prefix(path) {
+  const prefix = isRelative(path) ? '/assets/' : ''
```
Comment (brunoabreu) [inline react/utils/assets.js]: These `prefix` and `isRelative` are legacy code that does not make sense anymore. It also breaks the dynamic locales loading, take a look here https://github.com/vtex-apps/render-runtime/pull/11

Comment (firstdoit) [reply react/utils/assets.js]: Thanks! Will remove.

File: react/utils/events.js [hunk 2]
```diff
@@ -1,15 +1,17 @@
-import EventEmitter from 'events'
+import EventEmitter from 'eventemitter3'
 import {canUseDOM} from 'exenv'
 
 if (canUseDOM) {
-  global.__RUNTIME__.eventEmitter = global.__RUNTIME__.eventEmitter || new EventEmitter()
+  global.__RUNTIME__.emitter = global.__RUNTIME__.emitter || new EventEmitter()
+  // Backwards compatibility with old builder entrypoint code
```
[the remark this answers is not in the document]
Comment (firstdoit) [reply react/utils/events.js]: 😉

File: react/index.js [hunk 3]
```diff
@@ -60,37 +59,26 @@ const isRoot = (name, index, names) =>
...
-          pages={pages}
-          page={page}
-          settings={settings}
-          locale={locale}
-          messages={messages}
-        >
-          {component}
-        </RenderProvider>
-      </ApolloProvider>
+      <RenderProvider history={canUseDOM && isPage && !customRouting && createHistory()}>
```
[the remark this answers is not in the document]
Comment (firstdoit) [reply react/index.js]: I think it should be love ir or leave it. Do you suggest we control the URL when using `Link` components only? I think we should always try to control the URL, even if you have a simple href.

File: react/index.js [hunk 4]
```diff
@@ -24,18 +23,18 @@ if (global.IntlPolyfill) {
   }
 }
 
-const {account, culture: {locale}, messages, settings, extensions, pages, page} = global.__RUNTIME__
+const {culture: {locale}, extensions, pages, customRouting} = global.__RUNTIME__
```
[the remark this answers is not in the document]
Comment (firstdoit) [reply react/index.js]: OK, I'll put it inside the function that uses it directly. That should be enough.

**Comment by firstdoit** [comment]
@brunoabreu I think it's ready for a second look. 👁

File: react/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (brunoabreu) [reply react/index.js]: What is the purpose of the Link component? Why would anyone use it?

[the remark this answers is not in the document]
Comment (firstdoit) [reply react/index.js]: That's a great question 😂  do we need to manipulate history directly? Isn't navigating via href the same? So many questions

**Pull request merged**